### PR TITLE
Three qubit decomposition

### LINKED ISSUE #451: Implement general decomposition of 3-qubit unitary gates
This would allow ControlledGates made from 2-qubit KnownMatrixGates to automatically run on the simulator.

### PR DESCRIPTION
**Pull request by balopat**
Adds cirq.three_qubit_unitary_to_operations. 

For the design doc, please see the attached [3qubit-decomposition.pdf](https://github.com/quantumlib/Cirq/files/4996582/3qubit-decomposition.pdf)

Fixes #451.

### CODE REVIEW

**Review by Strilanc — DISMISSED** [review]
Note: Must wait until scipy release (can't merge the wheels)

**Comment by balopat** [comment]
Update from scipyDev mailing list: https://mail.python.org/pipermail/scipy-dev/2020-May/024154.html

Following schedule for 1.5.0 is accepted:
- May 26/2020: branch 1.5.x
- May 29/2020: rc1
- June 11/2020: rc2 (if needed)
- June 20/2020: final release

**Comment by balopat** [comment]
@viathor - the last one in the Three Qubit Decomposition Series :D Thank you for all the reviews so far!

File: cirq/optimizers/three_qubit_decomposition.py
Comment (viathor) [inline cirq/optimizers/three_qubit_decomposition.py:222]: This is a little awkward. For example, when we pass `shift=False` we ignore `d_w` so setting it to `np.eye(4)` is unnecessary (but the function signature forces us to return something...). It seems to me we can make it more elegant by moving the `if shift_left` conditional outside of this function. This also reduces the the number of arguments since it allows us to get rid of `shift_left`.

File: cirq/optimizers/three_qubit_decomposition.py
Comment (viathor) [inline cirq/optimizers/three_qubit_decomposition.py:158]: missing return type annotation

File: cirq/optimizers/three_qubit_decomposition.py
Comment (viathor) [inline cirq/optimizers/three_qubit_decomposition.py:165]: capitalize u->U, v->V to match D and W

(but let's keep small v in code)

Comment (viathor) [inline cirq/optimizers/three_qubit_decomposition.py:165]: Strictly speaking there should be parentheses around direct sums since matrix product takes priority.

(Probably moot due to another comment.)

Comment (viathor) [inline cirq/optimizers/three_qubit_decomposition.py:165]: All factors here are direct sums (i.e. block diagonal matrices), so this is a less readable variant of

$$
U_1 = V D W \\
U_2 = V D^\dagger W
$$

Comment (viathor) [inline cirq/optimizers/three_qubit_decomposition.py:165]: I think the latex will render better if we have underscore between letter and digit: U_1 etc.

File: cirq/optimizers/three_qubit_decomposition_test.py
Comment (viathor) [inline cirq/optimizers/three_qubit_decomposition_test.py:46]: We should also test that the number of CNOTs is at most 20.

File: cirq/optimizers/three_qubit_decomposition.py
Comment (balopat) [reply cirq/optimizers/three_qubit_decomposition.py:57]: Because we do not want to impose for this single functionality a lower bound on the full project until we are convinced that 1.5.0 is propagated to the majority of large, slower moving environments (e.g. ours internally, or national labs). I can create an issue for this and make a TODO for a year from now or so. 1.5.0 is very new, it just came out in July.

File: cirq/optimizers/three_qubit_decomposition.py
Comment (balopat) [reply cirq/optimizers/three_qubit_decomposition.py:244]: I think I'll push back on this one. I tried it and I'm not convinced - as while it might save some space, I'm not sure it will be more readable / understandable / testable. I really like the fact that the individual building blocks map to the concepts in the paper. I find that easier to follow. Also, the unit tests are thoroughly testing each primitive one by one. If I start to merge them further up, I'd have to write much more complicated tests.

File: cirq/optimizers/three_qubit_decomposition.py
Comment (balopat) [reply cirq/optimizers/three_qubit_decomposition.py:110]: Yeah a general compiler that does that could be useful, but much harder task than this specific one - that's why we agreed with Craig back in the day to implement the special case here. I can open an issue though, I think it might be useful. Though a lot of cases are covered by `MergeInteractions`, this might be an interesting one to write for larger gates. 

re: `cirq.linalg.is_involutive` could be cool - we just talked about with Orion maybe introducing a `cirq.period` protocol too, where things could tell their period, which would be 2 for involutary operators.

File: cirq/optimizers/three_qubit_decomposition.py
Comment (balopat) [reply cirq/optimizers/three_qubit_decomposition.py:69]: yes, with `[1, -1, 1, -1]`

File: cirq/optimizers/three_qubit_decomposition.py
Comment (balopat) [reply cirq/optimizers/three_qubit_decomposition.py:165]: Yes, that's right below it. I can remove it though if we don't think it's useful.

File: cirq/optimizers/three_qubit_decomposition.py
Comment (balopat) [reply cirq/optimizers/three_qubit_decomposition.py:222]: I know it's a bit awkward but theoretically correct and keeps the conceptual boundary nice and clean: this method returns the operations for a two-qubit multiplexor and extracts a diagonal optionally. That optionality is expressed via the diagonal being identity or not. 

If I pull out the if statement - then we have a method that "extracts some operations from the two-qubit multiplexor, and returns the W matrix for further processing"... It will mix the levels of abstraction in my opinion. 

I think I'll have to push back on this one as well, unless you see a resolution that is clean conceptually as well.

**Comment by balopat** [comment]
@viathor - ready for a final review - I hope you are okay with my reasoning around the two outstanding items.

File: cirq/optimizers/three_qubit_decomposition.py
Comment (viathor) [reply cirq/optimizers/three_qubit_decomposition.py:61]: SGTM

File: cirq/optimizers/three_qubit_decomposition.py
Comment (viathor) [reply cirq/optimizers/three_qubit_decomposition.py]: Fair enough.

Regarding tests: Unit tests that verify behavior of private help functions make refactoring harder. Ideally, tests would verify behavior of such helper functions via specially chosen input data. This is why many best practices docs (e.g. Google's own testing best practices) encourage testing public APIs.

File: cirq/optimizers/three_qubit_decomposition.py
Comment (viathor) [reply cirq/optimizers/three_qubit_decomposition.py]: OK

File: cirq/optimizers/three_qubit_decomposition.py
Comment (viathor) [reply cirq/optimizers/three_qubit_decomposition.py:226]: Sounds good - I agree about keeping clean mapping between code and concepts in the paper.

It still bothers me that we instantiate the identity matrix that the caller just ignores. Instantiation of this matrix creates an expectation in the mind of someone reading the code that it'll be used for something.

Suggestion: Let's use `None` in place of `np.eye(4)`. This has a nice ripple effect of changing the signature of the function by introducing explicit `Optional` that makes the optional character of the array explicit.

File: cirq/optimizers/three_qubit_decomposition.py
Comment (viathor) [inline cirq/optimizers/three_qubit_decomposition.py:162]: Let's make it `Tuple[Optional[np.ndarray], List[ops.Operation]]` per other comment.

File: cirq/optimizers/three_qubit_decomposition.py
Comment (viathor) [inline cirq/optimizers/three_qubit_decomposition.py:49]: We should probably use `atol` with `is_unitary`.

**Review by viathor — APPROVED** [review]
LGTM with a few minor comments.

File: cirq/optimizers/three_qubit_decomposition.py
Comment (balopat) [reply cirq/optimizers/three_qubit_decomposition.py:226]: I like that suggestion, thank you!

**Pull request merged**
